### PR TITLE
Add scoped profiling region

### LINKED ISSUE #4275: Add simple profiling class that pushes region on constructor and pops on destructor
To take advantage of the profiling hooks a programmer might write today
```
Kokkos::profiling::pushRegion("region_name");
...
Kokkos::profiling::popRegion();
```

An alternative model might be a one-line construct, like
```
Kokkos::profiling::profiler p("region_name");
```
This would be a simple class whose constructor calls `pushRegion` and whose destructor calls `popRegion`. Aside from saving overhead on typing, a limitation with the first model is that it gets tricky in functions with multiple returns, since you'd have to call `popRegion` just before every return. With the pop-on-destruction method there is no additional effort required for this case, we will automatically get a pop when the object goes out of scope.

### PR DESCRIPTION
**Pull request by aprokop**
Fix #4275

### CODE REVIEW

File: core/unit_test/tools/TestScopedProfileRegion.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  test_region_stack.push(label);
+}
+
+void test_pop_region() {
+  auto const &label = test_region_stack.top();
+  ASSERT_TRUE(true) << std::string("pop ") + label;
+  test_region_stack.pop();
+}
+
+}  // namespace
```
Comment (dalg24) [inline core/unit_test/tools/TestScopedProfileRegion.cpp:40]: Actually you should keep it open until the end of the file

File: core/unit_test/tools/TestScopedProfileRegion.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+#include <string>
+#include <stack>
+
+namespace {
+std::stack<std::string> test_region_stack;
+
+// NOTE: cannot use lambdas because they can only be converted to function
+// pointers if they don't capture anything
+void test_push_region(char const *label) {
+  ASSERT_TRUE(true) << std::string("push ") + label;
```
Comment (dalg24) [inline core/unit_test/tools/TestScopedProfileRegion.cpp:30]: I don't think it actually works.  I can't get the message to get printed.

**Review by dalg24 — COMMENTED** [review]
Looks good besides the status messages that don't get printed

File: core/unit_test/tools/TestScopedProfileRegion.cpp (hunk 2 above)
Comment (aprokop) [reply core/unit_test/tools/TestScopedProfileRegion.cpp:30]: Yeah, me neither.

File: core/unit_test/tools/TestScopedProfileRegion.cpp (hunk 1 above)
Comment (aprokop) [reply core/unit_test/tools/TestScopedProfileRegion.cpp:40]: I looked at the `TestProfilingSection.cpp`, which does it similarly.

File: core/src/Kokkos_Profiling_ScopedProfileRegion.hpp [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+#endif
+
+#include <Kokkos_Macros.hpp>
+#include <impl/Kokkos_Profiling.hpp>
+
+#include <string>
+
+namespace Kokkos::Profiling {
+
+class ScopedProfileRegion {
```
Comment (masterleinad) [inline core/src/Kokkos_Profiling_ScopedProfileRegion.hpp:31]:
```suggestion
class [[nodiscard]] ScopedProfileRegion {
```

File: core/src/Kokkos_Profiling_ScopedProfileRegion.hpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+ public:
+  ScopedProfileRegion(ScopedProfileRegion const &) = delete;
+  ScopedProfileRegion &operator=(ScopedProfileRegion const &) = delete;
+
+  explicit ScopedProfileRegion(std::string const &name) {
```
Comment (masterleinad) [inline core/src/Kokkos_Profiling_ScopedProfileRegion.hpp:36]:
```suggestion
#if defined(__has_cpp_attribute) && __has_cpp_attribute(nodiscard) >= 201907
  [[nodiscard]]
#endif
  explicit ScopedProfileRegion(std::string const &name) {
```

File: core/unit_test/tools/TestScopedProfileRegion.cpp (hunk 2 above)
Comment (dalg24) [reply core/unit_test/tools/TestScopedProfileRegion.cpp:30]: Yes.  Or if we want to see the info just use `printf("[          ] %s\n", ...)`

File: core/src/Kokkos_Profiling_ScopedProfileRegion.hpp (hunk 4 above)
Comment (masterleinad) [reply core/src/Kokkos_Profiling_ScopedProfileRegion.hpp:36]: See https://github.com/kokkos/kokkos/blob/a75aa23a1c54ff1afa7c13adf36bcee5694d4aea/core/src/Kokkos_Core.hpp#L221-L227 and https://github.com/kokkos/kokkos/pull/5224 to disallow creating a temporary object.

File: core/unit_test/tools/TestScopedProfileRegion.cpp (hunk 1 above)
Comment (dalg24) [reply core/unit_test/tools/TestScopedProfileRegion.cpp]: We have plenty of bad examples to copy from ;)

File: core/unit_test/tools/TestScopedProfileRegion.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+void test_pop_region() { test_region_stack.pop(); }
+
+TEST(defaultdevicetype, scoped_profile_region) {
+  Kokkos::Tools::Experimental::set_push_region_callback(test_push_region);
+  Kokkos::Tools::Experimental::set_pop_region_callback(test_pop_region);
+
+  ASSERT_TRUE(test_region_stack.empty());
+
+  // Unnamed guard!  Profile region is popped at the end of the statement.
+  Kokkos::Profiling::ScopedProfileRegion("bug");
```
Comment (aprokop) [inline core/unit_test/tools/TestScopedProfileRegion.cpp:41]: Because of `[[nodiscard]]`, this now throws a warning.

Comment (masterleinad) [reply core/unit_test/tools/TestScopedProfileRegion.cpp:41]: Yes, I would suggest removing it.

**Review by masterleinad — APPROVED** [review]
Looks OK to me.

**Review by dalg24 — APPROVED** [review]

**Comment by dalg24** [comment]
Ignoring SYCL failure that is clearly unrelated

**Comment by dalg24** [comment]
I realize now there was no discussion on the name `ScopedProfileRegion`.  Adding it to the agenda for the developer meeting tomorrow to give a chance to people who might have an opinion on this to weight in.

**Comment by dalg24** [comment]
> I realize now there was no discussion on the name `ScopedProfileRegion`. Adding it to the agenda for the developer meeting tomorrow to give a chance to people who might have an opinion on this to weight in.

Developer meeting decision: Rename to `ScopedRegion`

**Pull request merged**